File: Question_01_10/answers_cpp/answer_7.cpp
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <iostream>
#include <math.h>


// average pooling
cv::Mat average_pooling(cv::Mat img){
  int height = img.rows;
  int width = img.cols;
  int channel = img.channels();

  // prepare output
  cv::Mat out = cv::Mat::zeros(height, width, CV_8UC3);

  int r = 8;
  double v = 0;
  
  for (int y = 0; y < height; y+=r){
    for (int x = 0; x < width; x+=r){
      for (int c = 0; c < channel; c++){
        v = 0;
        for (int dy = 0; dy < r; dy++){
          for (int dx = 0; dx < r; dx++){
            v += (double)img.at<cv::Vec3b>(y + dy, x + dx)[c];
          }
        }
        v /= (r * r);
        for (int dy = 0; dy < r; dy++){
          for (int dx = 0; dx < r; dx++){
            out.at<cv::Vec3b>(y + dy, x + dx)[c] = (uchar)v;
          }
        }
      }
	  }
	}
  return out;
}

int main(int argc, const char* argv[]){
  // read image
  cv::Mat img = cv::imread("imori.jpg", cv::IMREAD_COLOR);

  // average pooling
  cv::Mat out = average_pooling(img);
  
  //cv::imwrite("out.jpg", out);
  cv::imshow("answer", out);
  cv::waitKey(0);
  cv::destroyAllWindows();

  return 0;
}

File: Question_01_10/answers_py/answer_7.py
import cv2
import numpy as np


# average pooling
def average_pooling(img, G=8):
    out = img.copy()

    H, W, C = img.shape
    Nh = int(H / G)
    Nw = int(W / G)

    for y in range(Nh):
        for x in range(Nw):
            for c in range(C):
                out[G*y:G*(y+1), G*x:G*(x+1), c] = np.mean(out[G*y:G*(y+1), G*x:G*(x+1), c]).astype(np.int)
    
    return out


# Read image
img = cv2.imread("imori.jpg")

# Average Pooling
out = average_pooling(img)

# Save result
cv2.imwrite("out.jpg", out)
cv2.imshow("result", out)
cv2.waitKey(0)
cv2.destroyAllWindows()
